foundations/src/telemetry/log/internal.rs

```diff
@@ -34,20 +35,39 @@ impl LoggerWithKvNestingTracking {
         }
     }
 
-    /// Increment the KV nesting level. You should call this every time you have replaced the
+    /// Increment the KV nesting level. You should call this before any time you're going to replace the
     /// logger with a child of itself.
     ///
-    /// If this returns true, you should first release any locks you may hold (they will be
-    /// poisoned) and then call the panic_from_too_much_nesting() associated function.
-    #[must_use]
-    pub(crate) fn has_too_much_nesting(&mut self) -> bool {
-        self.nesting_level += 1;
-        self.nesting_level > Self::MAX_NESTING
-    }
-
-    /// Please call this if has_too_much_nesting() returns true.
-    pub(crate) fn panic_from_too_much_nesting() -> ! {
-        panic!("{}", Self::EXCEEDED_MAX_NESTING_ERROR);
+    /// If this returns None, it will consume the logger lock, and you should not nest any further.
+    /// If panic_on_too_much_logger_nesting is enabled, instead of returning None this will free the
+    /// logger lock and then panic.
+    pub(crate) fn check_nesting_level(
+        mut current_log_lock: parking_lot::lock_api::RwLockWriteGuard<
+            parking_lot::RawRwLock,
+            LoggerWithKvNestingTracking,
+        >,
+    ) -> Option<
+        parking_lot::lock_api::RwLockWriteGuard<
+            parking_lot::RawRwLock,
+            LoggerWithKvNestingTracking,
+        >,
+    > {
+        current_log_lock.nesting_level += 1;
```

Comment:
We will eventually wrap around and start nesting again. I don't know how worried to be about this, but it might feel less icky if we incremented it only in the first and second arms of the match?

Reply:
I was being a bit optimistic that we wouldn't try to nest that much. I'll fix it.

Reply:
Or `saturating_add`

Reply:
saturating_add is good too!

Reply:
nice!